### ISSUE TITLE
[BUG] Holding on to a RMM allocation exception makes cuDF go OOM

### ISSUE THREAD

**Issue by madsbk** [rapidsai/cudf#9232]
**Describe the bug**
Trying to allocate a too large cuDF dataframe, RMM raises a  `std::bad_alloc` exception as expected. However, when holding on to the exception (a `MemoryError` exception in Python), subsequent _small_ dataframe allocations fails with the same `std::bad_alloc` exception. 
This issue relates to https://github.com/rapidsai/dask-cuda/issues/725

**Steps/Code to reproduce bug**
In the following code we try to allocate a small Series (`create_single_cudf_ser()`) **after** a big Series has failed (`create_big_cudf_ser()`).

Notice:
 - If not using a rmm pool (`pool_allocator=False`) it works
 - if we use `rmm.DeviceBuffer` instead of Series it works.
 - if we delete the exception before the next small allocation it works.
 - This could be RMM bug, but I was not able to reproduce it only using RMM.

```python
import cudf
import rmm

# Setting `pool_allocator=False` makes it work
rmm.reinitialize(pool_allocator=True)

# create_big_cudf_ser -> Causes OOM on a worker
def create_big_cudf_ser():
    # return rmm.DeviceBuffer(size=30*2**30) # <= Makes it work
    n_rows = 300_000_000
    s_1 = cudf.Series([1], dtype="int64").repeat(n_rows)
    s_2 = cudf.Series([1], dtype="int64").repeat(n_rows)
    s_3 = cudf.Series([1], dtype="int64").repeat(n_rows)
    s_4 = cudf.Series([1], dtype="int64").repeat(n_rows)
    s_5 = cudf.Series([1], dtype="int64").repeat(n_rows)
    s_6 = cudf.Series([1], dtype="int64").repeat(n_rows)
    s_7 = cudf.Series([1], dtype="int64").repeat(n_rows)
    return len(s_1) + len(s_2) + len(s_3) + len(s_4) + len(s_5) + len(s_6) + len(s_7)


# create_small_cudf_ser -> should succeed on a worker
def create_single_cudf_ser():
    # return rmm.DeviceBuffer(size=2**30)  # <= Makes it work
    n_rows = 200_000_000
    series_1 = cudf.Series([1], dtype="int64").repeat(n_rows)
    return len(series_1)


print("create_single_cudf_ser()")
create_single_cudf_ser()
err = None
try:
    print("create_big_cudf_ser()")
    create_big_cudf_ser()
except MemoryError as e:
    err = e
    print(e)
# del err  # <= Makes it work
print("create_single_cudf_ser()")
create_single_cudf_ser()
print("FINISHED")
```

**Comment by shwina** [comment]
@madsbk here's the output I'm getting when running your reproducer - it indicates that `create_single_cudf_ser()` _does_ finish without raising a `std::bad_alloc`:

```python
create_single_cudf_ser()
create_big_cudf_ser()
std::bad_alloc: CUDA error at: ../include/rmm/mr/device/cuda_memory_resource.hpp:69: cudaErrorMemoryAllocation out of memory
create_single_cudf_ser()
FINISHED
```

Can I ask if you're running this as a script, or on the REPL/Notebook?

**Comment by madsbk** [comment]
Yes, that looks correct. I am running it as a script and gets:
```
create_single_cudf_ser()
create_big_cudf_ser()
std::bad_alloc: RMM failure at:/datasets/mkristensen/miniconda3/envs/rap-0913/include/rmm/mr/device/pool_memory_resource.hpp:182: Maximum pool size exceeded
create_single_cudf_ser()
Traceback (most recent call last):
  File "rmm_omm_bug.py", line 40, in <module>
    create_single_cudf_ser()
  File "rmm_omm_bug.py", line 25, in create_single_cudf_ser
    series_1 = cudf.Series([1], dtype="int64").repeat(n_rows)
  File "/datasets/mkristensen/miniconda3/envs/rap-0913/lib/python3.8/site-packages/cudf/core/frame.py", line 1505, in repeat
    return self._repeat(repeats)
  File "/datasets/mkristensen/miniconda3/envs/rap-0913/lib/python3.8/site-packages/cudf/core/frame.py", line 1512, in _repeat
    *libcudf.filling.repeat(self, count)
  File "cudf/_lib/filling.pyx", line 58, in cudf._lib.filling.repeat
  File "cudf/_lib/filling.pyx", line 86, in cudf._lib.filling._repeat_via_size_type
MemoryError: std::bad_alloc: RMM failure at:/datasets/mkristensen/miniconda3/envs/rap-0913/include/rmm/mr/device/pool_memory_resource.hpp:182: Maximum pool size exceeded
```

Could you try running the following, which is the original Dask/Distributed issue:
```python
import cudf
import dask
from dask_cuda import LocalCUDACluster
from dask.distributed import Client
import distributed


# create_big_cudf_ser -> Causes OOM on a worker
def create_big_cudf_ser():
    n_rows = 300_000_000
    s_1 = cudf.Series([1], dtype="int64").repeat(n_rows)
    s_2 = cudf.Series([1], dtype="int64").repeat(n_rows)
    s_3 = cudf.Series([1], dtype="int64").repeat(n_rows)
    s_4 = cudf.Series([1], dtype="int64").repeat(n_rows)
    s_5 = cudf.Series([1], dtype="int64").repeat(n_rows)
    s_6 = cudf.Series([1], dtype="int64").repeat(n_rows)
    s_7 = cudf.Series([1], dtype="int64").repeat(n_rows)
    return len(s_1) + len(s_2) + len(s_3) + len(s_4) + len(s_5) + len(s_6) + len(s_7)


# create_small_cudf_ser -> should succeed on a worker
def create_single_cudf_ser():
    n_rows = 200_000_000
    series_1 = cudf.Series([1], dtype="int64").repeat(n_rows)
    return len(series_1)


def schedule_small_tasks():
    n_tasks = 1
    dask_small_array_tasks = [
        dask.delayed(create_single_cudf_ser)() for i in range(0, n_tasks)
    ]
    outputs = client.compute(dask_small_array_tasks)

    outputs = distributed.wait(outputs)
    completed_tasks = [o.result() for o in outputs.done]
    assert len(completed_tasks) == n_tasks
    print(completed_tasks[0:10])


def schedule_oom_tasks():
    n_tasks = 1
    dask_big_ser_tasks = [
        dask.delayed(create_big_cudf_ser)() for i in range(0, n_tasks)
    ]
    outputs = client.compute(dask_big_ser_tasks)
    outputs = distributed.wait(outputs)
    completed_tasks = [o.result() for o in outputs.done]
    assert len(completed_tasks) == n_tasks


if __name__ == "__main__":
    cluster = LocalCUDACluster(rmm_pool_size="28 GB", n_workers=1)
    # cluster = LocalCUDACluster(n_workers=1)
    client = Client(cluster)

    print("*" * 100)
    print("schedule_small_tasks()")
    schedule_small_tasks()  ### WORKS

    try:
        print("*" * 100)
        print("schedule_oom_tasks()")
        schedule_oom_tasks()  #### OOMS the worker
    except:
        pass

    print("*" * 100)
    print("schedule_small_tasks()")
    schedule_small_tasks()  #### Fails now as worker in bad state
    print("*" * 100)
    print("FINISHED")
```

**Comment by shwina** [comment]
Thanks! I get the same - investigating.

```
****************************************************************************************************
schedule_small_tasks()
[200000000]
****************************************************************************************************
schedule_oom_tasks()
distributed.worker - WARNING - Compute Failed
Function:  create_big_cudf_ser
args:      ()
kwargs:    {}
Exception: MemoryError('std::bad_alloc: RMM failure at:../include/rmm/mr/device/pool_memory_resource.hpp:179: Maximum pool size exceeded')

****************************************************************************************************
schedule_small_tasks()
distributed.worker - WARNING - Compute Failed
Function:  create_single_cudf_ser
args:      ()
kwargs:    {}
Exception: MemoryError('std::bad_alloc: RMM failure at:../include/rmm/mr/device/pool_memory_resource.hpp:179: Maximum pool size exceeded')

Traceback (most recent call last):
  File "rmm_test_dask.py", line 70, in <module>
    schedule_small_tasks()  #### Fails now as worker in bad state
  File "rmm_test_dask.py", line 36, in schedule_small_tasks
    completed_tasks = [o.result() for o in outputs.done]
  File "rmm_test_dask.py", line 36, in <listcomp>
    completed_tasks = [o.result() for o in outputs.done]
  File "/home/ashwin/workspace/rapids/compose/etc/conda/cuda_11.2/envs/rapids/lib/python3.7/site-packages/distributed/client.py", line 220, in result
    raise exc.with_traceback(tb)
  File "rmm_test_dask.py", line 24, in create_single_cudf_ser
    series_1 = cudf.Series([1], dtype="int64").repeat(n_rows)
  File "/home/ashwin/workspace/rapids/cudf/python/cudf/cudf/core/frame.py", line 1515, in repeat
    return self._repeat(repeats)
  File "/home/ashwin/workspace/rapids/cudf/python/cudf/cudf/core/frame.py", line 1522, in _repeat
    *libcudf.filling.repeat(self, count)
  File "cudf/_lib/filling.pyx", line 58, in cudf._lib.filling.repeat
  File "cudf/_lib/filling.pyx", line 86, in cudf._lib.filling._repeat_via_size_type
MemoryError: std::bad_alloc: RMM failure at:../include/rmm/mr/device/pool_memory_resource.hpp:179: Maximum pool size exceeded
```

**Comment by vyasr** [comment]
@madsbk @shwina @randerzander @VibhuJawa just wanted to give interested parties an update since I haven't been able to resolve this issue yet. Apologies in advance, this description will go a bit into the weeds but I think it's helpful to document my results so far in case someone else has ideas on how to proceed beyond what I'm doing.

I was able to reproduce the bug from the script in the initial issue, here's a slightly smaller MWE:

```
import cudf
import rmm

# Setting `pool_allocator=False` makes it work
rmm.reinitialize(pool_allocator=True)

def create_cudf_ser(n_rows, num):
    series_list = []
    for _ in range(num):
        # Works if I don't append to the list and instead create temporary
        # Series objects that are never saved
        series_list.append(cudf.Series([1], dtype="int64").repeat(n_rows))

print("Create one small series")
create_cudf_ser(200_000_000, 1)
err = None
try:
    print("Create two large series")
    create_cudf_ser(900_000_000, 2)  # Works if I instead call with (1_800_000_000, 1) or even larger
except MemoryError as e:
    err = e
    print(e)
# del err  # <= Makes it work
print("Create one small series")
create_cudf_ser(200_000_000, 1)
print("FINISHED")
```

The fact that deleting the exception made it work made me pretty suspicious of what was happening here, so I implemented `__del__` with a print statement for the Series, Column, and Buffer classes, and I added corresponding print statements in `__init__` to see what was happening (all `print(..., flush=True)` to ensure immediate output). Here's the output:

<details><summary>Without the `del err`</summary>
<p>

```
Create one small series
Creating series  140620888999312
Creating buffer  140618239789520
Creating column  140618053628560
Creating buffer  140618241784848
Creating column  140618239693312
Creating buffer  140618239816080
Creating column  140618238428224
Creating buffer  140618239789584
Creating column  140618053513440
Deleting series  140620888999312
Deleting column  140618053628560
Deleting buffer  140618239789520
Deleting column  140618239693312
Deleting buffer  140618241784848
Deleting series  140618239789968
Deleting column  140618053513440
Deleting buffer  140618239789584
Deleting column  140618238428224
Deleting buffer  140618239816080
Create two large series
Creating series  140618241784848
Creating buffer  140618239789968
Creating column  140618238428224
Creating buffer  140618239816272
Creating column  140618053513440
Creating buffer  140618239790224
Creating column  140618239693312
Creating buffer  140618239789712
Creating column  140618053628560
Deleting series  140618241784848
Deleting column  140618238428224
Deleting buffer  140618239789968
Deleting column  140618053513440
Deleting buffer  140618239816272
Creating series  140618241784848
Creating buffer  140618239789968
Creating column  140618053513440
Creating buffer  140618239816080
Creating column  140618238428224
std::bad_alloc: RMM failure at:../include/rmm/mr/device/pool_memory_resource.hpp:183: Maximum pool size exceeded
Create one small series
Creating series  140618238506192
Creating buffer  140618053464528
Creating column  140618053579552
Creating buffer  140618053464656
Creating column  140618053565040
Traceback (most recent call last):
  File "test2.py", line 24, in <module>
    create_cudf_ser(200_000_000, 1)
  File "test2.py", line 10, in create_cudf_ser
    series_list.append(cudf.Series([1], dtype="int64").repeat(n_rows))
  File "/home/nfs/vyasr/local/rapids/cudf/python/cudf/cudf/core/frame.py", line 1744, in repeat
    return self._repeat(repeats)
  File "/home/nfs/vyasr/local/rapids/cudf/python/cudf/cudf/core/frame.py", line 1751, in _repeat
    *libcudf.filling.repeat(self, count)
  File "cudf/_lib/filling.pyx", line 58, in cudf._lib.filling.repeat
  File "cudf/_lib/filling.pyx", line 86, in cudf._lib.filling._repeat_via_size_type
MemoryError: std::bad_alloc: RMM failure at:../include/rmm/mr/device/pool_memory_resource.hpp:183: Maximum pool size exceeded
Deleting series  140618238506192
Deleting column  140618053579552
Deleting buffer  140618053464528
Deleting column  140618053565040
Deleting buffer  140618053464656
Deleting series  140618241784848
Deleting series  140618238505936
Deleting column  140618238428224
Deleting column  140618053513440
Deleting buffer  140618239816080
Deleting column  140618053628560
Deleting buffer  140618239789968
Deleting buffer  140618239789712
Deleting column  140618239693312
Deleting buffer  140618239790224
```

</p>
</details>

<details><summary>With the `del err`</summary>
<p>

```
Create one small series
Creating series  140689729279504
Creating buffer  140686900742160
Creating column  140686894084752
Creating buffer  140687118690896
Creating column  140686901871104
Creating buffer  140687118689104
Creating column  140686900679744
Creating buffer  140686900741904
Creating column  140686893969632
Deleting series  140689729279504
Deleting column  140686894084752
Deleting buffer  140686900742160
Deleting column  140686901871104
Deleting buffer  140687118690896
Deleting series  140687675218512
Deleting column  140686893969632
Deleting buffer  140686900741904
Deleting column  140686900679744
Deleting buffer  140687118689104
Create two large series
Creating series  140687125825360
Creating buffer  140689729279504
Creating column  140686900679744
Creating buffer  140689729279696
Creating column  140686893969632
Creating buffer  140686893859856
Creating column  140686901871104
Creating buffer  140687118689104
Creating column  140686894084752
Deleting series  140687125825360
Deleting column  140686900679744
Deleting buffer  140689729279504
Deleting column  140686893969632
Deleting buffer  140689729279696
Creating series  140687125825360
Creating buffer  140689729279504
Creating column  140686893969632
Creating buffer  140689729279440
Creating column  140686900679744
std::bad_alloc: RMM failure at:../include/rmm/mr/device/pool_memory_resource.hpp:183: Maximum pool size exceeded
Deleting series  140687125825360
Deleting column  140686893969632
Deleting buffer  140689729279504
Deleting column  140686900679744
Deleting buffer  140689729279440
Deleting series  140686900741904
Deleting column  140686894084752
Deleting buffer  140687118689104
Deleting column  140686901871104
Deleting buffer  140686893859856
Create one small series
Creating series  140686893859856
Creating buffer  140687118689680
Creating column  140686901871104
Creating buffer  140687118690896
Creating column  140686894084752
Creating buffer  140687125825360
Creating column  140686900679744
Creating buffer  140687118689104
Creating column  140686893969632
Deleting series  140686893859856
Deleting column  140686901871104
Deleting buffer  140687118689680
Deleting column  140686894084752
Deleting buffer  140687118690896
Deleting series  140689729279696
Deleting column  140686893969632
Deleting buffer  140687118689104
Deleting column  140686900679744
Deleting buffer  140687125825360
FINISHED
```

</p>
</details>

As you can see, the number of objects that are deleted after the initial (caught) OOM exception and before the subsequent call to `create_cudf_ser` is larger if we delete the stored exception. If we don't delete the exception, those objects are only deleted at the very end after the uncaught MemoryError triggers a termination of the script and a full cleanup of the environment. The fact that deleting the error has this effect and some of the other corresponding fixes in the script above (not storing the series objects into local variables, making one larger object instead of multiple smaller ones) suggests a bug related to Python's reference counting. Manually forcing garbage collection inside or outside the function call does not appear to fix the issue, indicating that the most likely culprit is that the MemoryError object is actively maintaining a reference to the function-local stack frame and preventing its destruction.

It is possible that the problem is not in Python itself, but rather in Cython's handling of references when it propagates C++ exceptions to Python (in this case RMM's `std::bad_alloc`), but I am not yet sure of the best way to distinguish those since the changes required to make RMM fail silently instead of throw could leave its internal memory pool in an invalid internal state. I've inspected the Cython-generated C++ code and I'm not seeing anything immediately wrong with the macro it's using to propagate exceptions, but I'm not terribly familiar with Python's C APIs so it's possible that I'm missing something.

In summary, unless I'm missing something this appears to be a problem with Python itself, not with cudf or rmm (please correct me if you see an alternative). Whether this is coming from Python or Cython isn't clear, and I may have to try to build up a MWE independent of RAPIDS libraries to confirm that if I can't narrow this down by inspection, so I don't have a good idea of how much longer this will take to fix.

**Comment by shwina** [comment]
@madsbk for now, perhaps we could add a `del err` in the relevant part of the `dask-cuda` code (with a comment xref'ing this issue)?

**Comment by vyasr** [comment]
> @madsbk for now, perhaps we could add a `del err` in the relevant part of the `dask-cuda` code (with a comment xref'ing this issue)?

That would be my recommendation as well. Even if we can narrow this down it's likely that the fix will need to be upstreamed somewhere else and we wouldn't be able to bump the necessary dependencies any time soon.

**Comment by shwina** [comment]
Thanks @VibhuJawa -- I see the same (I had to uncomment the following line to see the error):

```
schedule_oom_tasks()  #### OOMS the worker
```

**Comment by VibhuJawa** [comment]
> Thanks @VibhuJawa -- I see the same (I had to uncomment the following line to see the error):
> 
> ```
> schedule_oom_tasks()  #### OOMS the worker
> ```

Thanks (edited in the code too).

**Comment by vyasr** [comment]
I can reproduce this as well.

**Comment by jrhemstad** [comment]
Anyone tried to repro in pure C++?

**Comment by vyasr** [comment]
> Anyone tried to repro in pure C++?

I have not, mostly because the evidence in my post above pretty clearly indicates that the problem is that the owning Python objects are simply not being destructed, so our C++ RAII design is moot. Moreover, I'm pretty confident that this is entirely unrelated to libcudf/librmm given some of the strange modifications to the Python script that allow the code to work, particularly when those should be exactly equivalent. One example is the list comprehension example above, where changing
```
def create_cudf_ser(n_rows, num):
    series_list = []
    for _ in range(num):
        series_list.append(cudf.Series([1], dtype="int64").repeat(n_rows))
```
to 
```
def create_cudf_ser(n_rows, num):
    series_list = [cudf.Series([1], dtype="int64").repeat(n_rows) for _ in range(num)]
```
makes it work. Another is that (with some additional changes to the script) I observe that
```
def create_cudf_ser(n_rows):
    series_list = []
    for _ in range(5):
        series_list.append(cudf.Series([1], dtype="int64").repeat(n_rows))
```
works, while 
```
def create_cudf_ser(n_rows):
    s_1 = cudf.Series([1], dtype="int64").repeat(n_rows)
    s_2 = cudf.Series([1], dtype="int64").repeat(n_rows)
    s_3 = cudf.Series([1], dtype="int64").repeat(n_rows)
    s_4 = cudf.Series([1], dtype="int64").repeat(n_rows)
    s_5 = cudf.Series([1], dtype="int64").repeat(n_rows)
```
fails. There's absolutely no reason for those to be different unless Python is failing to destroy objects whose reference counts reach zero. It seems likely that particular Python bytecodes are not playing nice with Cython's C++ exception forwarding mechanisms.

**Comment by VibhuJawa** [comment]
Inspired from @vyasr example, see a pure `cupy` example  , which has the same issue. 


```python
import cupy as cp

def create_cudf_ser(n_rows, num):
    s_1 = cp.asarray([1], dtype="int64").repeat(n_rows)
    s_2 = cp.asarray([1], dtype="int64").repeat(n_rows)
    s_3 = cp.asarray([1], dtype="int64").repeat(n_rows)
    s_4 = cp.asarray([1], dtype="int64").repeat(n_rows)
    s_5 = cp.asarray([1], dtype="int64").repeat(n_rows)
    

print("Create one small array")
create_cudf_ser(200_000_000, 1)
err = None
try:
    print("Create two large array")
    create_cudf_ser(900_000_000, 4)  # Works if I instead call with (1_800_000_000, 1) or even larger
except MemoryError as e:
    err = e
    print(e)

#del err  # <= Makes it work
print("Create one small series")
create_cudf_ser(400_000_000, 1)
print("FINISHED")
```

```python
---------------------------------------------------------------------------
OutOfMemoryError                          Traceback (most recent call last)
/tmp/ipykernel_70768/1937475258.py in <module>
     21 #del err  # <= Makes it work
     22 print("Create one small series")
---> 23 create_cudf_ser(400_000_000, 1)
     24 print("FINISHED")

/tmp/ipykernel_70768/1937475258.py in create_cudf_ser(n_rows, num)
      3 def create_cudf_ser(n_rows, num):
      4     s_1 = cp.asarray([1], dtype="int64").repeat(n_rows)
----> 5     s_2 = cp.asarray([1], dtype="int64").repeat(n_rows)
      6     s_3 = cp.asarray([1], dtype="int64").repeat(n_rows)
      7     s_4 = cp.asarray([1], dtype="int64").repeat(n_rows)

cupy/_core/core.pyx in cupy._core.core.ndarray.repeat()

cupy/_core/core.pyx in cupy._core.core.ndarray.repeat()

cupy/_core/_routines_manipulation.pyx in cupy._core._routines_manipulation._ndarray_repeat()

cupy/_core/_routines_manipulation.pyx in cupy._core._routines_manipulation._repeat()

cupy/_core/core.pyx in cupy._core.core.ndarray.__init__()

cupy/cuda/memory.pyx in cupy.cuda.memory.alloc()

cupy/cuda/memory.pyx in cupy.cuda.memory.MemoryPool.malloc()

cupy/cuda/memory.pyx in cupy.cuda.memory.MemoryPool.malloc()

cupy/cuda/memory.pyx in cupy.cuda.memory.SingleDeviceMemoryPool.malloc()

cupy/cuda/memory.pyx in cupy.cuda.memory.SingleDeviceMemoryPool._malloc()

cupy/cuda/memory.pyx in cupy.cuda.memory.SingleDeviceMemoryPool._try_malloc()

OutOfMemoryError: Out of memory allocating 3,200,000,000 bytes (allocated so far: 32,000,000,512 bytes).
```

**Comment by vyasr** [comment]
OK I have a true MWE here that I plan to post to the Cython Github. I'll follow up with what they have to say.

**Comment by vyasr** [comment]
Actually it turns out that while there is a minor Cython issue that I'll raise with them tomorrow, that's a much smaller problem than the main culprit, which can be demonstrated with pure Python. Here's a very simple script:

```
_MAX_COUNTER = 4
_COUNTER = 0


class RefCounter:
    def __init__(self):
        global _COUNTER, _MAX_COUNTER
        _COUNTER += 1
        if _COUNTER > _MAX_COUNTER:
            raise RuntimeError("Too many objects.")

    def __del__(self):
        global _COUNTER
        _COUNTER -= 1


def make_counters(num):
    counters = []
    for _ in range(num):
        counters.append(RefCounter())


counters1 = RefCounter()
print("The count is ", _COUNTER)

error = None
try:
    make_counters(5)
except RuntimeError as e:
    error = e
    print(e)

print("The count is ", _COUNTER)
# del error  # As before, we need to uncomment this line in order to avoid seeing failures.
print("The count is ", _COUNTER)

print("After")
counters3 = RefCounter()
print("Finished")
```

What we're running into is almost certainly that the frame corresponding to the level of the stack where the exception was originally raised is not being destroyed until the exception itself goes out of scope. You can trigger this sort of behavior explicitly [as described in the Python docs](https://docs.python.org/3/library/inspect.html#the-interpreter-stack), but I strongly suspect that simply keeping the exception around like this is sufficient to trigger the same behavior. I'll post a question on StackOverflow to see if there's an alternate solution, but I think what this ultimately comes down to is that if you hold on to the exception there are no guarantees about the function-local scope being destroyed.

**Comment by vyasr** [comment]
Looks like there is in fact [a StackOverflow post already discussing this](https://stackoverflow.com/questions/2055611/python-function-local-variable-scope-during-exceptions), along with a response from one of the core Python devs. I think what this boils down to is that it is not sufficient to wrap calls to the functions in our previous examples in try-except blocks. You _must_ wrap the constructions corresponding to allocations inside those blocks in a try-except and explicitly remove all references upon failure to guarantee that the desired deallocation occurs immediately rather than in a delayed fashion. So in my last example, you would want something like
```
def make_counters(num):
    try:
        counters = []
        for _ in range(num):
            counters.append(RefCounter())        
    except RuntimeError:
        counters = None
        raise

**Comment by vyasr** [comment]
In light of my last message, I'm going to close this issue as not something that we can control in cudf, but feel free to reopen if you feel like there's something more to be done here.

**Comment by madsbk** [comment]
@shwina, @vyasr, @VibhuJawa, thanks for the investigation. 
I think I have a general solution to the issue: https://github.com/dask/distributed/pull/5338
